imports/client/components/SetupPage.tsx

```diff
@@ -453,6 +461,119 @@ const GoogleDriveTemplateForm = ({ initialDocTemplate, initialSpreadsheetTemplat
+  error: Error;
+});
+
+const GoogleScriptForm = ({ app }: {
+  app: (SettingType & { name: 'google.script' })['value'] | undefined
+}) => {
+  const [state, setState] = useState<GoogleScriptFormState>({
+    submitState: SubmitState.IDLE,
+  });
+
+  const dismissAlert = useCallback(() => {
+    setState({ submitState: SubmitState.IDLE });
+  }, []);
+
+  const ensureScript = useCallback((e: React.MouseEvent<HTMLButtonElement>) => {
+    e.preventDefault();
+    setState({ submitState: SubmitState.SUBMITTING });
+    configureEnsureGoogleScript.call((error) => {
+      if (error) {
+        setState({ submitState: SubmitState.ERROR, error });
+      } else {
+        setState({ submitState: SubmitState.SUCCESS });
+      }
+    });
+  }, []);
+
+  const [endpointUrl, setEndpointUrl] = useState(app?.endpointUrl ?? '');
+  const onEndpointUrlChange = useCallback((e: React.ChangeEvent<HTMLInputElement>) => {
+    setEndpointUrl(e.currentTarget.value);
+  }, []);
+
+  const saveEndpointUrl = useCallback((e: React.MouseEvent<HTMLButtonElement>) => {
+    e.preventDefault();
+    if (!endpointUrl) return;
+
+    setState({ submitState: SubmitState.SUBMITTING });
+    configureGoogleScriptUrl.call({ url: endpointUrl }, (error) => {
+      if (error) {
+        setState({ submitState: SubmitState.ERROR, error });
+      } else {
+        setState({ submitState: SubmitState.SUCCESS });
+      }
+    });
+  }, [endpointUrl]);
+
+  const shouldDisable = state.submitState === 'submitting';
+  return (
+    <>
+      {state.submitState === 'submitting' ? <Alert variant="info">Saving...</Alert> : null}
+      {state.submitState === 'success' ? <Alert variant="success" dismissible onClose={dismissAlert}>Saved changes.</Alert> : null}
+      {state.submitState === 'error' ? (
+        <Alert variant="danger" dismissible onClose={dismissAlert}>
+          Saving failed:
+          {' '}
+          {state.error.message}
+        </Alert>
+      ) : null}
+      <p>
+        Jolly Roger uses Google Apps Script to access APIs which are
+        otherwise not accessible through Google&apos;s public APIs, such as
+        inserting images into a Google Spreadsheet.
+      </p>
+      <p>
+        We need to offer users a way to insert images, because the insert
+        image dialog within Google Spreadsheets prevents itself from working
+        within an iframe.
+      </p>
+      <p>
+        This is a two-step process. First we need to create the project in
+        Google Apps Script (or update the source code, if it is out of date):
+      </p>
+      <p>
+        <Button variant="primary" onClick={ensureScript} disabled={shouldDisable}>Create (or update) Google Apps Script project</Button>
+      </p>
+      {app?.scriptId && (
+        <>
+          <p>
+            Next, you need to create a deployment. Unfortunately, Google does not
+            expose this via API, so you&apos;ll need to do it by hand:
+          </p>
+          <ul>
```

Comment:
Suggest making this an `<ol>` instead -- the steps are ordered, and having numbers helps if you're jumping back and forth between browser windows.

Reply:
👍🏻